kore/src/Kore/Builtin/String.hs

```diff
@@ -310,14 +314,16 @@ evalString2Base = Builtin.applicationEvaluator evalString2Base0
             _str <- expectBuiltinString string2BaseKey _strTerm
             _base <- Int.expectBuiltinInt string2BaseKey _baseTerm
             packedResult <-
-                case _base of
-                    -- no builtin reader for number in octal notation
-                    8 -> return $ case readOct $ Text.unpack _str of
-                        [(result, "")] -> Right (result, "")
-                        _ -> Left ""
-                    10 -> return $ Text.signed Text.decimal _str
-                    16 -> return $ Text.signed Text.hexadecimal _str
-                    _ -> warnNotImplemented app >> empty
+              if 2 <= _base && _base <= 36
+                then let digitToIntM ch = elemIndex ch $ take (fromIntegral _base) "0123456789abcdefghijklmnopqrstuvwxyz"
+                         validDigit = isJust . digitToIntM
+                         digitToInt = fromMaybe 0 . digitToIntM
+                         lowerCaseStr = Text.unpack $ Text.toLower _str
+                         readWithBase = readSigned $ readInt _base validDigit digitToInt
+                     in return $ case readWithBase lowerCaseStr of
+                                   [(result, "")] -> Right (result, "")
+                                   _ -> Left ("" :: String)
+                else warnNotImplemented app >> empty
```

Comment:
Let's use `unless` and `empty` to avoid the body of the function floating right. This way, the "happy path" stays all together:

```suggestion
              unless (2 <= _base && _base <= 36) (warnNotImplemented app >> empty)
              let digitToIntM ch = elemIndex ch $ take (fromIntegral _base) "0123456789abcdefghijklmnopqrstuvwxyz"
                  validDigit = isJust . digitToIntM
                  digitToInt = fromMaybe 0 . digitToIntM
                  lowerCaseStr = Text.unpack $ Text.toLower _str
                  readWithBase = readSigned $ readInt _base validDigit digitToInt
              return $ case readWithBase lowerCaseStr of
                                   [(result, "")] -> Right (result, "")
                                   _ -> Left ("" :: String)
```

Let's also extract the pure part (`readWithBase`) to the top level. Finally, don't do indexing in a list, but get the digit values from the [`numeric representation`](https://hackage.haskell.org/package/base-4.15.0.0/docs/Data-Char.html#g:6) of `Char`.